### DISCUSSION TITLE
"Negative Jacobian" when using GapHeatTransfer

### DISCUSSION THREAD

**Discussion by WayneX-Wong** [idaholab/moose#27934] [Q&A Modules: General] [answered]
Hi there,

I was using GapHeatTransfer, but encountered a "negative jacobian" error.

**_Model:_**
A cylinder and a hollow cylinder with a 1cm gap between the bottom surface of the former and the top surface of the latter, GapHeatTransfer is set to this gap.
![image](https://github.com/idaholab/moose/assets/111510550/0d983a67-c6d9-4a17-9e26-e8c42378a706)

**_Input:_**
Shown as [main.txt](https://github.com/user-attachments/files/15894716/main.txt)

**_Error:_**
```
Elem Information
   id()=invalid, unique_id()=invalid, subdomain_id()=1, processor_id()=0
   type()=QUAD4
   dim()=2
   n_nodes()=4
   mapping=LAGRANGE_MAP
    0  Node id()=84581, processor_id()=0, Point=(x,y,z)=(-0.0962917, 0.0269797,      0.5)
    DoFs=(0/0/84581) (1/0/169162) (1/1/169163) 
    1  Node id()=84056, processor_id()=0, Point=(x,y,z)=(-0.112731,   0.0372,      0.5)
    DoFs=(0/0/84056) (1/0/168112) (1/1/168113) 
    2  Node id()=84451, processor_id()=0, Point=(x,y,z)=(-0.113464, 0.0171475,      0.5)
    DoFs=(0/0/84451) (1/0/168902) (1/1/168903) 
    3  Node id()=84580, processor_id()=0, Point=(x,y,z)=(-0.0990686, 0.0136167,      0.5)
    DoFs=(0/0/84580) (1/0/169160) (1/1/169161) 
   n_sides()=4
    neighbor(0)=nullptr
    neighbor(1)=nullptr
    neighbor(2)=nullptr
    neighbor(3)=nullptr
   hmin()=0.0136485, hmax()=0.027255
   volume()=0.000269658
   active()=1, ancestor()=0, subactive()=0, has_children()=0
   parent()=nullptr
   level()=0, p_level()=0
   refinement_flag()=DO_NOTHING
   p_refinement_flag()=DO_NOTHING
   DoFs=
libMesh terminating:
ERROR: negative Jacobian -4.1359e-25 at point (x,y,z)=(-0.0533052, 0.0001775,      0.5) in element 18446744073709551615
```
Negative Jacobian error occurs at the inner hollow on the top surface of the hollow cylinder, with no node exists there

**_Troubleshooting attempt:_**

1. I first checked the mesh volume and conformality using the MeshDiagnosticsGenerator and found no negative volume or non-confromal issues
```
*** Info ***
Number of elements below prescribed minimum volume : 0


*** Info ***
Number of elements with negative volume : 0


*** Info ***
Number of elements above prescribed maximum volume : 0


*** Info ***
Number of non-conformal nodes: 0
```
2. I then tried to change the hollow cylinder to a solid cylinder in the model, which was able to run normally without any error, and got a result which seems reasonable.

Is the problem due to the fact that GapHeatTransfer cannot applied to situation when two surfaces have different shapes?

Thanks!

**Comment by GiudGiud** [comment]
Hello

When using the MeshDiagnosticsGenerator in 1), did you use the element jacobian check?
If this reports an error, then your input mesh has a problem.

A close up to the elements around (-0.0962917, 0.0269797,      0.5)  (from the log) could show the issue

Guillaume

**Reply by WayneX-Wong** [reply]
Hello Guillaume

I use the element jacobian check, it reports no error
```
*** Info ***
Number of elements with a negative Jacobian: 0

*** Info ***
Number of element sides with negative Jacobians: 0
```

**Reply by GiudGiud** [reply]
Ok good. So the initial mesh is decent. 

The first step for the element deformation errors is to set the dirichlet BCs to "preset = false"

**Reply by WayneX-Wong** [reply]
I added "preset = false" in DirichletBC, but it didn't work, still same error.

**Reply by GiudGiud** [reply]
Next step would be to reduce the time step

**Reply by WayneX-Wong** [reply]
Actually it's a steady execution. My guess is whether GapHeatTransfer can only be used for heat transfer between regular surfaces without holes?

**Reply by GiudGiud** [reply]
Can you turn it into a transient and relax to steady state?
I m wondering if maybe the solids are expanding and you are getting inter-penetration between the two blocks

GapHeatTransfer requires a gap and two surfaces in front of each other, there is no radiation going through the large hole.
But this is not why the mesh is being deformed and the simulation crashing.

**Reply by WayneX-Wong** [reply]
I change my input into a transient one, and use very small time step, but it gives same error before solving.
```
Framework Information:
MOOSE Version:           git commit 82e31844cb on 2024-05-21
LibMesh Version:         
PETSc Version:           3.20.3
SLEPc Version:           3.20.1
Current Time:            Thu Jun 20 13:55:55 2024
Executable Timestamp:    Fri Jun  7 12:00:06 2024

Parallelism:
  Num Processors:          1
  Num Threads:             1

Mesh: 
  Parallel Type:           replicated
  Mesh Dimension:          3
  Spatial Dimension:       3
  Nodes:                   84600
  Elems:                   75648
  Num Subdomains:          2

Nonlinear System:
  Num DOFs:                84600
  Num Local DOFs:          84600
  Variables:               "T" 
  Finite Element Types:    "LAGRANGE" 
  Approximation Orders:    "FIRST" 

Auxiliary System:
  Num DOFs:                169200
  Num Local DOFs:          169200
  Variables:               { "penetration" "paired_T" } 
  Finite Element Types:    "LAGRANGE" 
  Approximation Orders:    "FIRST" 

Execution Information:
  Executioner:             Transient
  TimeStepper:             ConstantDT
  TimeIntegrator:          ImplicitEuler
  Solver Mode:             Preconditioned JFNK

  Elem Information
   id()=invalid, unique_id()=invalid, subdomain_id()=1, processor_id()=0
   type()=QUAD4
   dim()=2
   n_nodes()=4
   mapping=LAGRANGE_MAP
    0  Node id()=84581, processor_id()=0, Point=(x,y,z)=(-0.0962917, 0.0269797,      0.5)
    DoFs=(0/0/84581) (1/0/169162) (1/1/169163) 
    1  Node id()=84056, processor_id()=0, Point=(x,y,z)=(-0.112731,   0.0372,      0.5)
    DoFs=(0/0/84056) (1/0/168112) (1/1/168113) 
    2  Node id()=84451, processor_id()=0, Point=(x,y,z)=(-0.113464, 0.0171475,      0.5)
    DoFs=(0/0/84451) (1/0/168902) (1/1/168903) 
    3  Node id()=84580, processor_id()=0, Point=(x,y,z)=(-0.0990686, 0.0136167,      0.5)
    DoFs=(0/0/84580) (1/0/169160) (1/1/169161) 
   n_sides()=4
    neighbor(0)=nullptr
    neighbor(1)=nullptr
    neighbor(2)=nullptr
    neighbor(3)=nullptr
   hmin()=0.0136485, hmax()=0.027255
   volume()=0.000269658
   active()=1, ancestor()=0, subactive()=0, has_children()=0
   parent()=nullptr
   level()=0, p_level()=0
   refinement_flag()=DO_NOTHING
   p_refinement_flag()=DO_NOTHING
   DoFs=
libMesh terminating:
ERROR: negative Jacobian -4.1359e-25 at point (x,y,z)=(-0.0533052, 0.0001775,      0.5) in element 18446744073709551615
Stack frames: 21
0: libMesh::print_trace(std::ostream&)
1: libMesh::MacroFunctions::report_error(char const*, int, char const*, char const*, std::ostream&)
2: libMesh::FEMap::compute_single_point_map(unsigned int, std::vector<double, std::allocator<double> > const&, libMesh::Elem const*, unsigned int, std::vector<libMesh::Node const*, std::allocator<libMesh::Node const*> > const&, bool)
3: libMesh::FEMap::compute_map(unsigned int, std::vector<double, std::allocator<double> > const&, libMesh::Elem const*, bool)
4: libMesh::FE<2u, (libMesh::FEFamily)0>::reinit(libMesh::Elem const*, std::vector<libMesh::Point, std::allocator<libMesh::Point> > const*, std::vector<double, std::allocator<double> > const*)
5: Moose::findContactPoint(PenetrationInfo&, libMesh::FEGenericBase<double>*, libMesh::FEGenericBase<double>*, libMesh::FEType&, libMesh::Point const&, bool, double, bool&)
6: PenetrationThread::createInfoForElem(std::vector<PenetrationInfo*, std::allocator<PenetrationInfo*> >&, std::vector<PenetrationInfo*, std::allocator<PenetrationInfo*> >&, libMesh::Node const*, libMesh::Elem const*, std::vector<libMesh::Node const*, std::allocator<libMesh::Node const*> > const&, bool)
7: PenetrationThread::operator()(libMesh::StoredRange<__gnu_cxx::__normal_iterator<unsigned long*, std::vector<unsigned long, std::allocator<unsigned long> > >, unsigned long> const&)
8: /home/lagrange/projects/moose/framework/libmoose-opt.so.0(+0x1139f26) [0x7d98da139f26]
9: PenetrationLocator::detectPenetration()
10: GeometricSearchData::update(GeometricSearchData::GeometricSearchType)
11: FEProblemBase::updateGeomSearch(GeometricSearchData::GeometricSearchType)
12: FEProblemBase::initialSetup()
13: Transient::init()
14: MooseApp::executeExecutioner()
15: MooseApp::run()
16: ./pelican-opt(+0x30eb) [0x5557531e60eb]
17: main
18: /lib/x86_64-linux-gnu/libc.so.6(+0x29d90) [0x7d98cf829d90]
19: __libc_start_main
20: ./pelican-opt(+0x3396) [0x5557531e6396]
[0] ../src/fe/fe_map.C, line 903, compiled May 15 2024 at 11:58:00

application called MPI_Abort(MPI_COMM_WORLD, 1) - process 0
[unset]: write_line error; fd=-1 buf=:cmd=abort exitcode=1
:
system msg for write_line failure : Bad file descriptor
```

And I didn't add any ThermalExpansion related syntax, just a simple ThermalConduction problem, there shouldn't be solid expansion I think.

**Reply by maxnezdyur** [reply]
@GiudGiud 
In the error: ` ERROR: negative Jacobian -4.1359e-25 at point (x,y,z)=(-0.0533052, 0.0001775,      0.5) in element 18446744073709551615`, the element number seems way too high. Is this some sort of overflow problem or do we actually order elements like that?

**Reply by GiudGiud** [reply]
Good catch. We are reading out of bounds here. 

Can you share a MWE of this error? We'll take a look

**Reply by WayneX-Wong** [reply]
Sorry for the late reply, I put this input and mesh at [https://github.com/WayneX-Wong/GapTransTestMOOSE.git](url)

**Reply by GiudGiud** [reply]
Sorry for the delay I have not had time.

In the mean time I would try a full mesh instead of having a hole, and simply de-activate elements in the center of one part.
You can do that by placing them in a separate block and block-restricting all variables outside of this block

**Reply by WayneX-Wong** [reply]
@GiudGiud 
Thanks for your suggestion! I will try

**Reply by WayneX-Wong** [reply]
I created a block, filled in the hole, and defined another variable and NullKernel in this block, other input remain unchanged.

But still same negative jacobian error with high element number.

**Reply by maxnezdyur** [reply]
Can you post the input file with the mesh. The link above is not accessible.

**Reply by GiudGiud** [reply]
the link isnt set up well
copy paste it in the browser

**Reply by maxnezdyur** [reply]
It fails in `fe_side->reinit(side, &points);` below (FindContactPoint.C).  I don't think this has anything to do with going out of bounds. For this example we are 4 iterations in and then it fails. I think we just need to have a try and catch here or something like that. If it can't converge then maybe it wasn't a good contact point? But I am very confused about how findContactPoint works.
``` // Newton Loop
  for (; nit < 12 && update_size > TOLERANCE * TOLERANCE; nit++)
  {
    d = secondary_point - phys_point[0];

    DenseMatrix<Real> jac(dim - 1, dim - 1);

    jac(0, 0) = (d2xyz_dxi2[0] * d) - (dxyz_dxi[0] * dxyz_dxi[0]);

    if (dim - 1 == 2)
    {
      jac(1, 0) = (d2xyz_dxieta[0] * d) - (dxyz_dxi[0] * dxyz_deta[0]);

      jac(0, 1) = (d2xyz_detaxi[0] * d) - (dxyz_deta[0] * dxyz_dxi[0]);
      jac(1, 1) = (d2xyz_deta2[0] * d) - (dxyz_deta[0] * dxyz_deta[0]);
    }

    DenseVector<Real> rhs(dim - 1);

    rhs(0) = -dxyz_dxi[0] * d;

    if (dim - 1 == 2)
      rhs(1) = -dxyz_deta[0] * d;

    DenseVector<Real> update(dim - 1);

    jac.lu_solve(rhs, update);

    ref_point(0) += update(0);

    if (dim - 1 == 2)
      ref_point(1) += update(1);

    points[0] = ref_point;
    fe_side->reinit(side, &points);
    d = secondary_point - phys_point[0];

    update_size = update.l2_norm();
  }
```

**Reply by GiudGiud** [reply, marked as answer]
@maxnezdyur you were on the right track

With a try/catch we can get further
@WayneX-Wong please check out this branch and see if the results are good for you
https://github.com/idaholab/moose/pull/28149

It still has to skip some contact nodes because it cant find a solution. But at least the solve starts

**Reply by maxnezdyur** [reply]
Oh nice, happy to help. @BoZeng1997 I believe you had some of these problem many months ago. I think the solution is finally found! Look at this thread and check #28149.

**Reply by WayneX-Wong** [reply]
The case runs successfully, and the results seem reasonable now!

Thanks again for your help @GiudGiud @maxnezdyur !
